server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -0,0 +1,191 @@
+                }
+            }
+        }
+
+        return createIndicator(status.getStatus(), status.getSummary(), status.getDetails());
+    }
+
+    private static class ShardAllocationCounts {
+        private int unassigned = 0;
+        private int unassigned_restarting = 0;
+        private int initializing = 0;
+        private int started = 0;
+
+        public void increment(ShardRouting routing, NodesShutdownMetadata metadata) {
+            switch (routing.state()) {
+                case UNASSIGNED -> {
+                    if (isUnassignedDueToTimelyRestart(routing, metadata)) {
+                        unassigned_restarting++;
+                    } else {
+                        unassigned++;
+                    }
+                }
+                case INITIALIZING -> initializing++;
+                case STARTED -> started++;
+            }
+        }
+    }
+
+    private static boolean isUnassignedDueToTimelyRestart(ShardRouting routing, NodesShutdownMetadata metadata) {
+        var info = routing.unassignedInfo();
+        if (info == null || info.getReason() != UnassignedInfo.Reason.NODE_RESTARTING) {
+            return false;
+        }
+        var shutdown = metadata.getAllNodeMetadataMap().get(info.getLastAllocatedNodeId());
+        if (shutdown == null || shutdown.getType() != SingleNodeShutdownMetadata.Type.RESTART) {
+            return false;
+        }
+        var now = System.currentTimeMillis();
+        var restartingAllocationDelayExpiration = info.getUnassignedTimeInMillis() + shutdown.getAllocationDelay().getMillis();
+        return now <= restartingAllocationDelayExpiration;
+    }
+
+    private static class ShardAllocationStatus {
+        private final ShardAllocationCounts primaries = new ShardAllocationCounts();
+        private final ShardAllocationCounts replicas = new ShardAllocationCounts();
+
+        public void addPrimary(ShardRouting routing, NodesShutdownMetadata metadata) {
+            primaries.increment(routing, metadata);
+        }
+
+        public void addReplica(ShardRouting routing, NodesShutdownMetadata metadata) {
+            replicas.increment(routing, metadata);
+        }
+
+        public HealthStatus getStatus() {
+            if (primaries.unassigned > 0) {
+                return RED;
+            } else if (replicas.unassigned > 0) {
+                return YELLOW;
+            } else {
+                return GREEN;
+            }
+        }
+
+        public String getSummary() {
+            var builder = new StringBuilder("This cluster has ");
+            if (primaries.unassigned > 0
+                || primaries.unassigned_restarting > 0
+                || replicas.unassigned > 0
+                || replicas.unassigned_restarting > 0) {
+                builder.append(
+                    Stream.of(
+                        createMessage(primaries.unassigned, "unavailable primary", " unavailable primaries"),
```

Comment:
nit: did you consider implement a `getSummary()` method in ShardAllocationCounts directly, and call the methods here?

Reply:
Yes, but I did not come up with a good way to merge distinct messages. It would require appending primary/primaries/replica/replicas suffixes as well as properly setting comas.